### PR TITLE
webpsave: add progress hook

### PR DESCRIPTION
**Pull request by dloebl**
libwebp supports a progress callback which reports the current encoding progress in percent: https://github.com/webmproject/libwebp/blob/36a6eea3bc2871c7f0166193a6fa42241af4a464/src/webp/encode.h#L342
```
// Progress hook, called from time to time to report progress. It can return
// false to request an abort of the encoding process, or true otherwise if
// everything is OK.
typedef int (*WebPProgressHook)(int percent, const WebPPicture* picture);
```
Abort encoding if the image has been marked as killed.
Our use case for this is that we use [vips_image_set_kill()](https://www.libvips.org/API/current/VipsImage.html#vips-image-set-kill) for our timeout handling, however this doesn't work well with `webpsave` at the moment, because `webpsave` passes the whole image data at once to libwebp.

### CODE REVIEW

**Comment by dloebl** [comment]
**Note to the reviewer:** 
`vips_image_eval( in,  VIPS_IMAGE_N_PELS( in ) );` feels hacky, but I'm not sure if there is a better/existing way of reporting progress during encoding.

**Review by kleisauke — COMMENTED** [review]
Interesting, I left a comment inline.

I assume this is an issue when saving a large non-animated WebP image? AFAIK, `vips_image_iskilled()` is checked by `vips_sink_disc()`, but that's only once per strip.

**Comment by dloebl** [comment]
> I assume this is an issue when saving a large non-animated WebP image?

Yes exactly. For large WebP saves we "hang" in libwebp for quite a while - without any progress reported.

**Example:**
Timeout `webpsave` after 5 seconds of processing a large input.

_webptest.c_
```
#include <stdio.h>
#include <vips/vips.h>

static void progress_callback(VipsImage *image, VipsProgress *progress, void *pdata) {
  if(progress->run >= 5) {
    vips_image_set_kill(image, TRUE);
  }
}

int main() {
  VipsImage *in = vips_image_new_from_file("test.png", NULL);
  g_signal_connect(in, "eval", G_CALLBACK(progress_callback), NULL);
  vips_webpsave(in, "out.webp", NULL);
  printf("%s", vips_error_buffer());
  return 0;
}
```

```
$ vips gaussnoise test.png 16000 16000
$ gcc -o webptest $(pkg-config --cflags --libs vips) webptest.c
```

**Current [master](02213d4031ac580d4fdb3ccce1306155657aeedf):**
```
$ time ./webptest
./webptest  113.24s user 3.69s system 100% cpu 1:56.53 total
```

**With [dloebl:webpsave-add-progress-hook](https://github.com/libvips/libvips/pull/3333):**
```
$ time ./webptest              
VipsImage: killed for image "test.png"
webpsave: unable to encode
webpsave: mux error
./webptest  6.29s user 1.05s system 116% cpu 6.314 total
```

File: libvips/foreign/webpsave.c [hunk 1]
```diff
@@ -104,6 +104,7 @@ typedef struct _VipsForeignSaveWebp {
 	 */
 	VipsForeignSaveWebpMode mode;
 	VipsImage *image;
+	VipsImage *in;
```
[the remark this answers is not in the document]
Comment (dloebl) [reply libvips/foreign/webpsave.c]: Good point! 👍
Removed the unnecessary extra variable with 9ee14b0.

**Comment by lovell** [comment]
Thanks Daniel, is it possible to add a test for this, perhaps using the callback/timeout from your example? We'd need to watch out for the problem where an expected-failure test might erroneously succeed when run on faster hardware.

**Comment by dloebl** [comment]
@lovell
>  Is it possible to add a test for this, perhaps using the callback/timeout from your example?

Good point. I've added a test with e0624bc.

**Comment by dloebl** [comment]
Something I'm a bit worried about, is that this PR causes the `eval` callbacks to be called after the `posteval` callbacks.. Not ideal I would guess.. :/

```
[..]
EVAL: percent: 99
EVAL: percent: 100
EVAL: percent: 100
EVAL: percent: 100
EVAL: percent: 100
EVAL: percent: 100
POSTEVAL: percent: 100
EVAL: percent: 100
EVAL: percent: 100
[..]
```

Any complaints on emitting a new signal (like `"encode"`) which reports actual encoding progress?

**Comment by kleisauke** [comment]
> Something I'm a bit worried about, is that this PR causes the `eval` callbacks to be called after the `posteval` callbacks.

Do you have a reproducer for this? I think that should not happen since the `posteval` callback is only executed _after_ `vips_sink_disc()` is done.
https://github.com/libvips/libvips/blob/65355b98bfab043b225f50203553b8aedab1ddbf/libvips/iofuncs/sinkdisc.c#L534

Perhaps you forgot to call `vips_image_set_progress( image, TRUE );`?

FWIW, I was unable to reproduce this on the https://github.com/dloebl/libvips/compare/webpsave-add-progress-hook...kleisauke:webpsave-add-progress-hook changeset using this test program:
https://gist.github.com/kleisauke/35623b85a63dd65e3ce2abecdfa7dacf

**Comment by dloebl** [comment]
@kleisauke 
> Perhaps you forgot to call vips_image_set_progress( image, TRUE );?

Looks like that's indeed the issue. I've reverted the` "encoding"` signal again.
It's working as expected when calling `vips_image_set_progress( image, TRUE );` beforehand.

**Comment by dloebl** [comment]
Thanks Kleis! Your changes look good to me 👍. Feel free to push to the branch.

**Comment by kleisauke** [comment]
Backport (+ changelog entry) to [`8.14`](https://github.com/libvips/libvips/tree/8.14) available as #3353.

**Pull request merged**